File: backend/app/services/compare_pillars.py
from __future__ import annotations

from typing import Any

from app.licensing import verify_license_sync
from app.services.scoring_input import is_skipped_status, unpack_inputs_payload

_DEFAULTS: dict[str, float | int] = {
    "pol": 72.0,
    "bdiv": 40.0,
    "ceo": 0,
    "whistleblower": 3.0,
    "audit_quality": 3.0,
    "ai_governance": 3.0,
    "transparency": 3.0,
    "explainability": 3.0,
    "data_ethics": 3.0,
    "model_robustness": 3.0,
    "algo_bias": 2.0,
    "safety_testing": 3.0,
    "data_privacy": 3.0,
    "cybersecurity": 3.0,
    "data_quality": 3.0,
    "regulatory_compliance": 3.0,
    "egy": 3.0,
    "circ": 3.0,
    "ghg": 3.0,
    "biodiversity": 3.0,
    "renewable_pct": 3.0,
    "ghg_intensity": 3.0,
    "wat": 2.0,
    "crm": 0,
}


def _skipped(status: dict[str, str], key: str) -> bool:
    return is_skipped_status(status.get(key))


def _f(v: dict[str, Any], key: str) -> float:
    d = _DEFAULTS.get(key, 3.0)
    x = v.get(key, d)
    if x is None:
        return float(d)
    return float(x)


def _i(v: dict[str, Any], key: str) -> int:
    d = int(_DEFAULTS.get(key, 0))
    x = v.get(key, d)
    if x is None:
        return d
    return int(x)


def _slider_0_100(v: dict[str, Any], status: dict[str, str], key: str) -> float | None:
    if _skipped(status, key):
        return None
    val = _f(v, key)
    return (val - 1.0) / 4.0 * 100.0


def _slider_inv_0_100(v: dict[str, Any], status: dict[str, str], key: str) -> float | None:
    if _skipped(status, key):
        return None
    val = _f(v, key)
    return (5.0 - val) / 4.0 * 100.0


def _mean(parts: list[float]) -> float | None:
    if not parts:
        return None
    return sum(parts) / len(parts)


def compute_pillar_scores(raw_inputs: dict[str, Any]) -> dict[str, float | None]:
    if not verify_license_sync():
        return {"governance": None, "transparency": None, "privacy": None, "environmental": None}
    values, status = unpack_inputs_payload(raw_inputs)
    v = values

    gov_parts: list[float] = []
    if not _skipped(status, "pol"):
        gov_parts.append(_f(v, "pol"))
    if not _skipped(status, "bdiv"):
        gov_parts.append(_f(v, "bdiv"))
    if not _skipped(status, "ceo"):
        gov_parts.append((1.0 - float(_i(v, "ceo"))) * 100.0)
    for k in ("whistleblower", "audit_quality", "ai_governance"):
        p = _slider_0_100(v, status, k)
        if p is not None:
            gov_parts.append(p)

    trans_parts: list[float] = []
    for k in ("transparency", "explainability", "data_ethics", "model_robustness", "safety_testing"):
        p = _slider_0_100(v, status, k)
        if p is not None:
            trans_parts.append(p)
    p = _slider_inv_0_100(v, status, "algo_bias")
    if p is not None:
        trans_parts.append(p)

    priv_parts: list[float] = []
    for k in ("data_privacy", "cybersecurity", "data_quality", "regulatory_compliance"):
        p = _slider_0_100(v, status, k)
        if p is not None:
            priv_parts.append(p)

    env_parts: list[float] = []
    for k in ("egy", "circ", "ghg", "biodiversity", "renewable_pct", "ghg_intensity"):
        p = _slider_0_100(v, status, k)
        if p is not None:
            env_parts.append(p)
    p = _slider_inv_0_100(v, status, "wat")
    if p is not None:
        env_parts.append(p)
    if not _skipped(status, "crm"):
        env_parts.append((1.0 - float(_i(v, "crm"))) * 100.0)

    return {
        "governance": _mean(gov_parts),
        "transparency": _mean(trans_parts),
        "privacy": _mean(priv_parts),
        "environmental": _mean(env_parts),
    }

File: lib/compare-pillars.ts
/**
 * Four-axis pillar scores (0–100) for compare / benchmark views.
 *
 * Keep in sync with: projects/invest-nl/backend/app/services/compare_pillars.py
 */

export type PillarScores = {
  governance: number | null;
  transparency: number | null;
  privacy: number | null;
  environmental: number | null;
};

const STATUS_NA = "na";
const STATUS_NO_INFO = "no_info";

export function isSkippedStatus(st: string | null | undefined): boolean {
  return st === STATUS_NA || st === STATUS_NO_INFO;
}

export function unpackInputsPayload(raw: Record<string, unknown>): {
  values: Record<string, unknown>;
  factorStatus: Record<string, string>;
} {
  const vals = raw.values;
  if (vals && typeof vals === "object" && !Array.isArray(vals)) {
    const status = raw.factor_status;
    const outStatus: Record<string, string> = {};
    if (status && typeof status === "object" && !Array.isArray(status)) {
      for (const [k, v] of Object.entries(status)) {
        if (typeof k === "string") outStatus[k] = String(v);
      }
    }
    return { values: { ...(vals as Record<string, unknown>) }, factorStatus: outStatus };
  }
  return { values: { ...raw }, factorStatus: {} };
}

const DEFAULTS: Record<string, number> = {
  pol: 72,
  bdiv: 40,
  ceo: 0,
  whistleblower: 3,
  audit_quality: 3,
  ai_governance: 3,
  transparency: 3,
  explainability: 3,
  data_ethics: 3,
  model_robustness: 3,
  algo_bias: 2,
  safety_testing: 3,
  data_privacy: 3,
  cybersecurity: 3,
  data_quality: 3,
  regulatory_compliance: 3,
  egy: 3,
  circ: 3,
  ghg: 3,
  biodiversity: 3,
  renewable_pct: 3,
  ghg_intensity: 3,
  wat: 2,
  crm: 0,
};

function skipped(status: Record<string, string>, key: string): boolean {
  return isSkippedStatus(status[key]);
}

function f(v: Record<string, unknown>, key: string): number {
  const d = DEFAULTS[key] ?? 3;
  const x = v[key];
  if (x === null || x === undefined) return d;
  return Number(x);
}

function i(v: Record<string, unknown>, key: string): number {
  const d = Math.trunc(DEFAULTS[key] ?? 0);
  const x = v[key];
  if (x === null || x === undefined) return d;
  return Math.trunc(Number(x));
}

function slider0_100(v: Record<string, unknown>, status: Record<string, string>, key: string): number | null {
  if (skipped(status, key)) return null;
  const val = f(v, key);
  return ((val - 1) / 4) * 100;
}

function sliderInv0_100(v: Record<string, unknown>, status: Record<string, string>, key: string): number | null {
  if (skipped(status, key)) return null;
  const val = f(v, key);
  return ((5 - val) / 4) * 100;
}

function mean(parts: number[]): number | null {
  if (!parts.length) return null;
  return parts.reduce((a, b) => a + b, 0) / parts.length;
}

/** Same thresholds as backend `score_to_grade` (scoring.py). */
export function scoreToGrade(s: number | null | undefined): string | null {
  if (s == null || Number.isNaN(s)) return null;
  const x = Math.round(s);
  if (x >= 90) return "A+";
  if (x >= 80) return "A";
  if (x >= 72) return "B+";
  if (x >= 62) return "B";
  if (x >= 52) return "C+";
  if (x >= 42) return "C";
  if (x >= 30) return "D";
  return "F";
}

export function pillarBandLabel(score: number | null | undefined): string | null {
  if (score == null || Number.isNaN(score)) return null;
  if (score >= 70) return "Strong";
  if (score >= 45) return "Moderate";
  return "Weak";
}

const PILLAR_LABELS: Record<keyof PillarScores, string> = {
  governance: "Governance",
  transparency: "Transparency",
  privacy: "Privacy",
  environmental: "Environmental",
};

export function strengthGapFromPillars(p: PillarScores): { strength: string | null; gap: string | null } {
  const entries = (Object.keys(PILLAR_LABELS) as (keyof PillarScores)[])
    .map((k) => ({ k, v: p[k] }))
    .filter((x): x is { k: keyof PillarScores; v: number } => x.v != null && !Number.isNaN(x.v));
  if (entries.length < 2) return { strength: null, gap: null };
  let maxK = entries[0].k;
  let minK = entries[0].k;
  let maxV = entries[0].v;
  let minV = entries[0].v;
  for (const e of entries) {
    if (e.v > maxV) {
      maxV = e.v;
      maxK = e.k;
    }
    if (e.v < minV) {
      minV = e.v;
      minK = e.k;
    }
  }
  return {
    strength: PILLAR_LABELS[maxK],
    gap: PILLAR_LABELS[minK],
  };
}

export function compositeGradeFromPillars(p: PillarScores): string | null {
  const vals = [p.governance, p.transparency, p.privacy, p.environmental].filter(
    (x): x is number => x != null && !Number.isNaN(x),
  );
  if (!vals.length) return null;
  const m = vals.reduce((a, b) => a + b, 0) / vals.length;
  return scoreToGrade(m);
}

export function computePillarScores(rawInputs: Record<string, unknown> | null | undefined): PillarScores {
  const raw = rawInputs && typeof rawInputs === "object" ? rawInputs : {};
  const { values: v, factorStatus: status } = unpackInputsPayload(raw);

  const govParts: number[] = [];
  if (!skipped(status, "pol")) govParts.push(f(v, "pol"));
  if (!skipped(status, "bdiv")) govParts.push(f(v, "bdiv"));
  if (!skipped(status, "ceo")) govParts.push((1 - i(v, "ceo")) * 100);
  for (const k of ["whistleblower", "audit_quality", "ai_governance"] as const) {
    const p = slider0_100(v, status, k);
    if (p != null) govParts.push(p);
  }

  const transParts: number[] = [];
  for (const k of ["transparency", "explainability", "data_ethics", "model_robustness", "safety_testing"] as const) {
    const p = slider0_100(v, status, k);
    if (p != null) transParts.push(p);
  }
  {
    const p = sliderInv0_100(v, status, "algo_bias");
    if (p != null) transParts.push(p);
  }

  const privParts: number[] = [];
  for (const k of ["data_privacy", "cybersecurity", "data_quality", "regulatory_compliance"] as const) {
    const p = slider0_100(v, status, k);
    if (p != null) privParts.push(p);
  }

  const envParts: number[] = [];
  for (const k of ["egy", "circ", "ghg", "biodiversity", "renewable_pct", "ghg_intensity"] as const) {
    const p = slider0_100(v, status, k);
    if (p != null) envParts.push(p);
  }
  {
    const p = sliderInv0_100(v, status, "wat");
    if (p != null) envParts.push(p);
  }
  if (!skipped(status, "crm")) envParts.push((1 - i(v, "crm")) * 100);

  return {
    governance: mean(govParts),
    transparency: mean(transParts),
    privacy: mean(privParts),
    environmental: mean(envParts),
  };
}
